carma_wm/src/IndexedDistanceMap.h

```diff
@@ -0,0 +1,195 @@
+#pragma once
+
+/*
+ * Copyright (C) 2019 LEIDOS.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+#include <vector>
+#include <tuple>
+#include <utility>
+#include <lanelet2_core/primitives/LineString.h>
+#include <lanelet2_core/geometry/Point.h>
+#include <unordered_map>
+#include <math.h>
+
+namespace carma_wm
+{
+/*!
+ * \brief O(1) distance lookup structure for quickly accessing route distance information
+ *
+ * This class is meant to be used when a route update occurs to precompute distances along the route to support rapid
+ * queries later on Insertion is O(n) where n is the number of points in the linestring This structure does not support
+ * element reinsertion Linestrings added to this structure should be fully unique (NO DUPLICATE IDs for linestring or
+ * point objects)
+ *
+ *
+ * NOTE: Pre-computing route distances make queries much faster but could slow down route loading for large routes and
+ * cause them to use more memory.
+ */
+class IndexedDistanceMap
+{
+private:
+  // Distance storage structure
+  // A vector of tuples where the first vector is indexed by line segment and the second vector is index by point index
+  // on that line segment. The double value in the tuple stores the total along-line distance to the start of that line
+  // segment from the first point on the first line segment The double value inside the interior vector store the
+  // along-line distance of the point from the start of its line segment
+  std::vector<std::tuple<std::vector<double>, double>> accum_lengths;
+
+  // Id mapping structure
+  // Sores the linestring and point index's as values with their lanelet Ids as the key
+  std::unordered_map<lanelet::Id, std::pair<size_t, size_t>> id_index_map;
+
+public:
+  /*!
+   * \brief Add a linestring to this structure. This function will iterate over the line string to compute distances
+   * between each segment.
+   *
+   * \throws std::invalid_argument If the provided linstring was already added to this structure
+   *
+   * \param ls The linestring to add
+   */
+  void pushBack(const lanelet::LineString2d& ls)
```

Comment:
Given the minimal functionality in 90% of the functions I figured a header file was easier since it was internal to the library. I can change it if you want.

Reply:
Since it's internal, it's probably fine. Was just curious what the reason was.